[jit] Make torch::utils::Future and ivalue::future apis closer

jjlilley opened the pull request:
This change harmonizes some aspects of the api.
  - torch::utils::Future callback should have no args, like ivalue::future.
     Many of the lines of this change are related to fixing that up downstream.

     No args makes the api simpler to use, particularly since many/most of the
     downstream use cases ignore the passed-in args. It's simple enough to
     appropriately capture the future in the lambda if necessary.

   - Add error/hasError methods to ivalue::Future.
   - Use c10::optional underneath for error to ivalue::Future.
   - Change markCompleted(error) to setError(error) to ivalue::Future.
   - Add setValue(FutureError) version to torch::utils::Future

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

Commit:
[WIP] Steps to make ivalue::future api closer to torch::utils::Future

This change harmonizes some aspects of the api. There fortunately aren't yet too
many use cases to fix up downstream:
 - Change the callback api args to match that of torch::utils::Future
 - Add error/hasError methods
 - Use c10::optional underneath for error
 - Change markCompleted(error) to setError(error)

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

Commit:
Update on "[WIP] Steps to make ivalue::future api closer to torch::utils::Future"


This change harmonizes some aspects of the api. There fortunately aren't yet too
many use cases to fix up downstream:
 - Change the callback api args to match that of torch::utils::Future
 - Add error/hasError methods
 - Use c10::optional underneath for error
 - Change markCompleted(error) to setError(error)

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

Commit:
Update on "[WIP] Steps to make ivalue::future api closer to torch::utils::Future"


This change harmonizes some aspects of the api. There fortunately aren't yet too
many use cases to fix up downstream:
 - Change the callback api args to match that of torch::utils::Future
 - Add error/hasError methods
 - Use c10::optional underneath for error
 - Change markCompleted(error) to setError(error)

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

Commit:
Update on "[WIP] Steps to make ivalue::future api closer to torch::utils::Future"


This change harmonizes some aspects of the api. There fortunately aren't yet too
many use cases to fix up downstream:
 - Change the callback api args to match that of torch::utils::Future
 - Add error/hasError methods
 - Use c10::optional underneath for error
 - Change markCompleted(error) to setError(error)

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

wanchaol reviewed (commented):
Some questions on the api, I am wondering should we do the other way around for the callback api (honoring ivalue callback api instead).

jjlilley commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -235,10 +235,12 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     std::string error_msg;
   };
+  using Callback =
+    std::function<void(const IValue&, const c10::optional<FutureError>&)>;
```
My thought here was to make the callback pattern effectively the same as utils::Future. This is one of the bigger divergences in the api.

That said, thinking about it more, I agree that the no-args format is more intuitive and easier to justify.

Capturing the future in the case where we care about the result is straightforward enough, and meanwhile cases like interpreter.cpp that don't care about the value aren't burdened by the wider api.

I'll try to amend and see is this looks like a better approach.

jjlilley replied:
@wanchaol fwiw, after having written this up (which I'll upload in a few more minutes), I agree with you even more here now.

Most of the downstream use-cases don't even use the args, and the impl just feels a lot cleaner without these unused args all over the place.

Commit:
Update on "[jit] Steps to make ivalue::future api closer to torch::utils::Future"


This change harmonizes some aspects of the api. There fortunately aren't yet too
many use cases to fix up downstream:
 - Change the callback api args to match that of torch::utils::Future
 - Add error/hasError methods
 - Use c10::optional underneath for error
 - Change markCompleted(error) to setError(error)

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

Commit:
Update on "[jit] Make torch::utils::Future and ivalue::future apis closer"

This change harmonizes some aspects of the api.
  - torch::utils::Future callback should have no args, like ivalue::future.
     Many of the lines of this change are related to fixing that up downstream.

     No args makes the api simpler to use, particularly since many/most of the
     downstream use cases ignore the passed-in args. It's simple enough to
     appropriately capture the future in the lambda if necessary.

   - Add error/hasError methods to ivalue::Future.
   - Use c10::optional underneath for error to ivalue::Future.
   - Change markCompleted(error) to setError(error) to ivalue::Future.
   - Add setValue(FutureError) version to torch::utils::Future

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

Commit:
fix dist_engine-related issue on "[wip] Make torch::utils::Future and ivalue::future apis closer"

This change harmonizes some aspects of the api.
  - torch::utils::Future callback should have no args, like ivalue::future.
     Many of the lines of this change are related to fixing that up downstream.

     No args makes the api simpler to use, particularly since many/most of the
     downstream use cases ignore the passed-in args. It's simple enough to
     appropriately capture the future in the lambda if necessary.

   - Add error/hasError methods to ivalue::Future.
   - Use c10::optional underneath for error to ivalue::Future.
   - Change markCompleted(error) to setError(error) to ivalue::Future.
   - Add setValue(FutureError) version to torch::utils::Future

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

jjlilley commented:
fwiw, I think this change should be ready for a look.

wanchaol reviewed (approved):
Looks good! pretty neat with captures! some nits and questions inline.

jjlilley commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -326,15 +323,25 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
-  std::mutex mutex_;
+  void runCbs(std::unique_lock<std::mutex>& lock) {
+    std::vector<std::function<void(void)>> cbs;
+    cbs.swap(callbacks_);
```
Two goals:
 - we don't want to hold the lock when running the callback. That can lead to deadlocks (e.g. if the callback calls value() it would deadlock)
 - we need to clear callbacks_ after they're invoked.

Swapping with an empty vector does this. 

The other approach is
  std::vector<std::function<void(void)>> cbs{std::move(callbacks_)};

But the move constructor doesn't explicitly guarantee callbacks_ will be empty after (even though it will be in real life).

jjlilley commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -379,7 +371,7 @@ void DistEngine::cleanupBackwardPass(const ContextPtr& autogradContext) {
   const auto& futureGrads =
       autogradContext->retrieveGraphTask()->future_result_;
-  TORCH_INTERNAL_ASSERT(futureGrads.use_count() == 1);
+  TORCH_INTERNAL_ASSERT(futureGrads.use_count() <= 2);
```
No, there is a throw-away test case where this doesn't happen. Part of me wanted to get rid of the check altogether, since asserting the number of references in non-test-case can be brittle.

jjlilley commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1344,12 +1344,12 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
 
   void handleError(const ExceptionMessage& msg, bool is_jit_exception) {
-    std::stringstream ss;
+    std::ostringstream ss;
```
Either is fine, it's a super minor micro-opt.

stringstream is effectively both an input and output stringstream. If you only need half the functionality, the constructor is cheaper for ostringstream. Because of c++ locale requirements, the constructor is a bit more expensive than you might expect

jjlilley commented on torch/csrc/utils/future.h:
```diff
@@ -28,17 +28,13 @@ class TORCH_API FutureError final : public std::exception {
 
   const T& wait() {
     std::unique_lock<std::mutex> lock(mutex_);
     finished_cv_.wait(lock, [this] { return completed_.load(); });
```
No, wait() doesn't complete the value.

jjlilley commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -91,8 +91,9 @@ void DistAutogradContext::accumulateGrad(
       new_grad,
-      // Add +1 here since we can't std::move(grad) when call AccumulateGrad::callHooks,
-      // since it is a const ref, and that incurs a refcount bump for the new_grad.
+      // Add +1 here since we can't std::move(grad) when call
```
I think 'arc lint' did this automatically for me? I'll undo it.

Commit:
Update on "[jit] Make torch::utils::Future and ivalue::future apis closer"

This change harmonizes some aspects of the api.
  - torch::utils::Future callback should have no args, like ivalue::future.
     Many of the lines of this change are related to fixing that up downstream.

     No args makes the api simpler to use, particularly since many/most of the
     downstream use cases ignore the passed-in args. It's simple enough to
     appropriately capture the future in the lambda if necessary.

   - Add error/hasError methods to ivalue::Future.
   - Use c10::optional underneath for error to ivalue::Future.
   - Change markCompleted(error) to setError(error) to ivalue::Future.
   - Add setValue(FutureError) version to torch::utils::Future

Differential Revision: [D20803251](https://our.internmc.facebook.com/intern/diff/D20803251/)

[ghstack-poisoned]

jjlilley commented:
Thanks for the comments. I just uploaded a version that
 - fixed the comment nit
 - also effectively rolled back a few lines of moving of code to runCbs() in ivalue_inl.h (this felt slightly unnecessary moving code around in retrospect).

malfet commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -326,15 +339,14 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
 
  private:
-  std::mutex mutex_;
+  mutable std::mutex mutex_;
```
Just curious: why mutex should be declared as mutable here?

malfet replied:
Sorry, misread `mutable` as `volatile` here.

jjlilley replied:
Right, it's so I can make error()/etc functions const

xush6528 commented on torch/csrc/utils/future.h:
```diff
@@ -109,41 +114,40 @@ class TORCH_API Future final {
     if (completed_) {
       lock.unlock();
-      callback(value_, error_);
-      return;
```
@jjlilley Do we really want to remove this return?

jjlilley replied:
Good catch, thank you!

The good news is the impact should be relatively minimal (it would exacerbate memory leaks case by reference cycles), but definitely a bug.

Fix at: https://github.com/pytorch/pytorch/pull/36662
